src/split_at.rs

```diff
@@ -233,59 +213,611 @@ unsafe impl<T> SplitAt for [T] {
+    ///
+    /// `source` is a pointer to a slice DST; see also `Self::l_len`.
+    source: T,
+    /// The length of the future left half of `source`.
+    ///
+    /// # Safety
+    ///
+    /// `l_len` is no greater than `source`'s length.
+    l_len: usize,
+}
+
+impl<T> Split<T> {
+    /// Produces a `Split` of `source` with `l_len`.
+    ///
+    /// # Safety
+    ///
+    /// `l_len` is no greater than `source`'s length.
+    #[inline(always)]
+    unsafe fn new(source: T, l_len: usize) -> Self {
+        Self { source, l_len }
+    }
+}
+
+impl<'a, T> Split<&'a T>
 where
-    I: invariant::Invariants,
-    T: ?Sized + pointer::Read<I::Aliasing, R> + SplitAt,
+    T: ?Sized + SplitAt,
 {
-    let inner = ptr.as_inner();
-
-    // SAFETY: The caller promises that `l_len.get()` is not greater than the
-    // length of `self`'s trailing slice.
-    let (left, right) = unsafe { inner.split_at_unchecked(l_len) };
-
-    // Lemma 0: `left` and `right` conform to the aliasing invariant
-    // `I::Aliasing`. Proof: If `I::Aliasing` is `Exclusive` or `T` permits
-    // interior mutation, the caller promises that `l_len.padding_needed_for()
-    // == 0`. Consequently, by post-condition on `PtrInner::split_at_unchecked`,
-    // there is no trailing padding after `left`'s final element that would
-    // overlap into `right`. If `I::Aliasing` is shared and `T` forbids interior
-    // mutation, then overlap between their referents is permissible.
-
-    // SAFETY:
-    // 0. `left` conforms to the aliasing invariant of `I::Aliasing, by Lemma 0.
-    // 1. `left` conforms to the alignment invariant of `I::Alignment, because
-    //    the referents of `left` and `Self` have the same address and type
-    //    (and, thus, alignment requirement).
-    // 2. `left` conforms to the validity invariant of `I::Validity`, neither
-    //    the type nor bytes of `left`'s referent have been changed.
-    let left = unsafe { Ptr::from_inner(left) };
-
-    // SAFETY:
-    // 0. `right` conforms to the aliasing invariant of `I::Aliasing, by Lemma
-    //    0.
-    // 1. `right` conforms to the alignment invariant of `I::Alignment, because
-    //    if `ptr` with `I::Alignment = Aligned`, then by invariant on `T:
-    //    SplitAt`, the trailing slice of `ptr` (from which `right` is derived)
-    //    will also be well-aligned.
-    // 2. `right` conforms to the validity invariant of `I::Validity`, because
-    //    `right: [T::Elem]` is derived from the trailing slice of `ptr`, which,
-    //    by contract on `T: SplitAt::Elem`, has type `[T::Elem]`.
-    let right = unsafe { Ptr::from_inner(right) };
-
-    (left, right)
+    #[inline(always)]
+    fn into_ptr(self) -> Split<Ptr<'a, T, (Shared, Aligned, Valid)>> {
+        // SAFETY: `Ptr::from_ref(self.source)` points to exactly `self.source`
+        // and thus maintains the invariants of `self` with respect to `l_len`.
+        unsafe { Split::new(Ptr::from_ref(self.source), self.l_len) }
```

Comment:
Assign `Ptr::from_ref(self.source)` to a variable outside of the `unsafe` block so it's clear what the `unsafe` block is guarding?

Reply:
https://github.com/google/zerocopy/compare/b21d06c408367bc92de75f99017fa67dad9e8aa4..187c1794711c6487fe424506c9132cf9b92f8576#diff-394d6458153031257865a7aaddcda3ac55a4ae61c642dee241136b08a66fcf90R259-R262